File: src/_141/Solution.java
package _141;

import structure.ListNode;

/**
 * Definition for singly-linked list.
 * class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) {
 * val = x;
 * next = null;
 * }
 * }
 *
 * @author relish
 * @since 2018/11/15
 */
public class Solution {

    public boolean hasCycle(ListNode head) {
        ListNode fast = head, slow = head;
        if (fast == null) return false;
        while (slow != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                return true;
            }
            if (fast == null) break;
        }
        return false;
    }

    public static void main(String[] args) {
        System.out.println(new Solution().hasCycle(ListNode.createTestData("[1,2]")));
    }
}

File: src/_141/kotlin/Solution.kt
package _141.kotlin

import structure.ListNode


/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun hasCycle(head: ListNode?): Boolean {
        var slow = head
        var fast = head
        while (slow != null && fast!!.next != null) {
            slow = slow.next
            fast = fast.next?.next
            if (fast == null) return false
            if (slow == fast) return true
        }
        return false
    }
}


fun main(args: Array<String>) {

}